checker/tests/nullness/ConditionalPolyNull.java

```diff
@@ -0,0 +1,11 @@
+import org.checkerframework.checker.nullness.qual.PolyNull;
+
+class ConditionalPolyNull {
+    public static @PolyNull String toLowerCaseA(@PolyNull String text) {
+        return text == null ? null : text.toLowerCase();
+    }
+
+    public static @PolyNull String toLowerCaseB(@PolyNull String text) {
+        return text != null ? text.toLowerCase() : null;
```

Comment:
Could you add the two combinations that should raise errors, to make sure we don't break that direction?

Reply:
Thanks for filing #532.
I'm not sure how you mean... in this version: `return text == null ? text.toLowerCase() : null ;` when text is non-null, the return value is null, so it the returned value violates the `@PolyNull` contract. Does that make sense?

Reply:
Thanks for the explanation. Yes, this makes sense. Then how to understand the error message below?
```
found : @FBCBottom @PolyNull String
required : @Initialized @NonNull String
```
The required type is `@NonNull` make sense. Why the found type is `@PolyNull` for return value 'null'?

Reply:
I think you're talking about this error message:

````
checker/tests/nullness/ConditionalPolyNull.java:21: error: [method.invocation.invalid] call to toLowerCase() not allowed on the given receiver.
        return text != null ? null : text.toLowerCase();
                                                     ^
  found   : @FBCBottom @PolyNull String
  required: @Initialized @NonNull String
````

For which you already filed the separate issue #532.
Is that the error you mean? If so, we should discuss in #532. If not, please clarify which error you're talking about.

Reply:
Sorry, I miscopied the wrong error message. This is the error I am referencing: 
```
checker/tests/nullness/ConditionalPolyNull.java:16: error: [return.type.incompatible] incompatible types in return.
        return text == null ? text.toLowerCase() : null ;
                            ^
  type of expression: @Initialized @Nullable String
  method return type: @Initialized @PolyNull String
```
If `text == null` condition is false, the `@PolyNull` of both method return type and argument type will be conservatively replaced with `@NonNull` by data flow refinement. However, the method return type is still `@PolyNull` here. 

Should the `@PolyNull` be replaced by `@NonNull` here?

Reply:
But after replacing `@PolyNull` with `@NonNull` in the return type, you would still get a mismatch - the returned expression is `@Nullable`.
Let's try to fix the type changes in `commonAssignmentCheck` and move them to somewhere more appropriate. Then we can see whether the types make sense or not.

Reply:
Yes, you are right. :)

Reply:
I believe these two are actually  related #540, because when I add `@PolyNull` to https://github.com/eisop/checker-framework/blob/d1d2d48f68f3da081df82d6edfa0f200e19cb319/checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java#L688C101-L688C101, the odd error `method.invocation.invalid` is gone.

Reply:
Yes, that's why we split of #540 from this issue.